### PR TITLE
Windows build

### PR DESCRIPTION
**Pull request by ocefpaf**
At some point this recipe worked on `staged-recipes`. I guess that some of the updates on the dependencies broke it.

@gillins and @msarahan do you have any idea where I should start? It seems that `zlib` is the issue here, but I am not sure how to fix it.

### CODE REVIEW

**Comment by msarahan** [comment]
It might be assuming that z.lib is a static library. You might need to add some option or preprocessor flag to tell it that is is a link library instead.

**Comment by gillins** [comment]
Do we know why it is looking under `C:\aroot` for zlib? Seems very strange... I just had a look at FindZLIB.cmake and saw this:
```
# Hints
# ^^^^^
#
# A user may set ``ZLIB_ROOT`` to a zlib installation root to tell this
# module where to look.
```
Perhaps set add `-D ZLIB_ROOT=%LIBRARY_PREFIX%`? I should have more time to look at this tomorrrow.

**Comment by msarahan** [comment]
C:\aroot is a holdover from the continuum internal build system. That indicates an incomplete recipe conversion (bug)

File: recipe/bld.bat [hunk 1]
```diff
@@ -0,0 +1,19 @@
+mkdir %SRC_DIR%\build
+cd %SRC_DIR%\build
+
+cmake -G "NMake Makefiles" ^
+      -D CMAKE_INSTALL_PREFIX=%LIBRARY_PREFIX% ^
+      -D BUILD_SHARED_LIBS=ON ^
+      -D ENABLE_TESTS=OFF ^
+      -D CMAKE_PREFIX_PATH=%LIBRARY_PREFIX% ^
+      -D ZLIB_LIBRARY=%LIBRARY_LIB%\zlib.lib ^
```
Comment (patricksnape) [inline recipe/bld.bat]: This is static linking - is that correct?

Comment (msarahan) [inline recipe/bld.bat]: What tells you that it is static?

Comment (patricksnape) [inline recipe/bld.bat]: The `.lib`? I thought in Windows `.lib` implied static and `.dll` implied dynamic, much as `.so` and `.a` on Linux for dynamic/static?

Comment (ocefpaf) [inline recipe/bld.bat]: @patricksnape I don't know the first thing about Windows. Sorry, but I am driving blind here. I also tried setting `ZLIB_ROOT` with no avail.

**Comment by patricksnape** [comment]
I'll have to pencil this in to check tomorrow when I'm back on my Windows PC - on my laptop at the moment so not much use. Might be that explicitly setting the `%LIBRARY_BIN%\zlib.dll` rather than `%LIBRARY_LIB%\zlib.lib` helps.

File: recipe/bld.bat (hunk 1 above)
Comment (msarahan) [inline recipe/bld.bat]: Nope. .lib can be either. Only way to know for sure is to use tools on the  lib. You can't directly link against a dll at build time.

Comment (gillins) [inline recipe/bld.bat]: z.lib, zdll.lib and zlib.lib are all import libraries for zlib.dll. zlibstatic.lib is the static library.

**Comment by gillins** [comment]
OK I think I have found it. In hdf5-targets.cmake (under %LIBRARY_PREFIX%\cmake) there is the following bit of code:
```
set_target_properties(hdf5 PROPERTIES
  INTERFACE_INCLUDE_DIRECTORIES "${_IMPORT_PREFIX}/include"
  INTERFACE_LINK_LIBRARIES "C:/aroot/stage/Library/lib/z.lib;C:/aroot/stage/Library/lib/z.lib"
)
```
This file was installed by hdf5 and is designed to be picked up by other packages building with cmake that use hdf5 and set things automatically - cmake is fairly complex I'm afraid. Seems that wherever hdf5 was built had a zlib under c:\aroot that got picked up instead of the conda one...

**Comment by patricksnape** [comment]
Great find @gillins.

**Comment by ocefpaf** [comment]
Let's try our latest `hdf5` then.

**Comment by msarahan** [comment]
Indeed, good find @gillins. That comes from the legacy continuum build
system.

Historical side note: many of these low level libraries used to be a random
hodgepodge of static and dynamic (generally whatever was the default in any
build example). I have been working to build both static and dynamic
libraries, and I have been keeping the dynamic library as the one that is
found by default. The static one should have some extra filename addition,
like you see here.

If you  get old releases from defaults, you might get static libraries with
the "good" name, and no dll.  Keep an eye out.

**Comment by jakirkham** [comment]
`hdf5-targets.cmake` so we might have a similar issue here, but it may still pass anyways. We should check if we need to tell `conda-build` to fix the paths in `hdf5-targets.cmake`.

**Comment by ocefpaf** [comment]
@jakirkham it seems that pinning to `hdf5 1.8.16` fixes the issue. I wanted a build with the old `hdf5` to avoid triggering the full rebuild of the stack, but I guess don't will not be possible. Here are the options:

- we do not merge this just yet so we can have at least 1 meeting before changing the pinning;
- we merge this and start updating the pinning  right away.

**Comment by jakirkham** [comment]
How about option 3, which I planned on doing anyways. We build our own version of `hdf5` 1.8.15.1, but bump the build number so we can steer clear of this break.

**Comment by ocefpaf** [comment]
> How about option 3

Sounds good. I will work on the new pinning once I get back home (June 6th).

**Comment by jakirkham** [comment]
> I will work on the new pinning once I get back home (June 6th)

Oh that's right you have back-to-back conferences. Geez. When did you have time to write this?

**Comment by ocefpaf** [comment]
Sitting down having beers with my mom while I wait for my flight to Portland :wink:

**Comment by pelson** [comment]
So is this good to go @ocefpaf?

**Comment by ocefpaf** [comment]
> So is this good to go @ocefpaf?

No yet. We need to pin `hdf5` to `1.8.16*` first, otherwise we will break all packages that install `libnetcdf` and `hdf5` together (like `netcdf4-python`).

I will be back in office next week and I can start doing that.

**Comment by jakirkham** [comment]
I tried to do builds of HDF5 1.8.15 and 1.8.15.1, but they are missing stuff on Windows. If someone could please investigate this that would be much appreciated. This would save us from having to break with `defaults`.

If we do decide to break with `defaults`, let's skip to 1.8.17 instead.

**Comment by jakirkham** [comment]
FYI HDF5 1.8.17 is being released at this moment.

**Comment by ocefpaf** [comment]
> If we do decide to break with defaults, let's skip to 1.8.17 instead.

:+1:

**Comment by jakirkham** [comment]
Restarted Travis CI and CircleCI here as they started before the HDF5 builds on the feedstock were done. They should now pass.

The AppVeyor build of `libnetcdf` is later in the queue than HDF5 so we should be good on that front.

**Comment by jakirkham** [comment]
Still need to pin the following for HDF5:

* [ ] `gdal`
* [ ] `libradolan`
* [ ] `bob.io.base`
* [ ] `vigra`
* [ ] `netcdf4`
* [ ] `libmatio`

Still need to pin the following for libnetcdf:

* [ ] `cdat-lite`
* [ ] `gdal`
* [ ] `cdo`
* [ ] `vigra`
* [ ] `netcdf4`
* [ ] `eccodes`
* [ ] `ecmwf_grib`

**Comment by ocefpaf** [comment]
Those that are failing should be fixed once this one gets merged.

**Comment by jakirkham** [comment]
Some of them should be. Some of them only depend on HDF5 and/or were failing previously before this pinning bump.

**Comment by ocefpaf** [comment]
> Some of them should be. Some of them only depend on HDF5 and/or were failing previously before this pinning bump.

Can you point me to those here in this issue? I can only see failures due to version mismatch that should be fixed once this PR gets merged.

**Comment by ocefpaf** [comment]
Ping @conda-forge/libradolan, @conda-forge/bob.io.base, @conda-forge/vigra, and @conda-forge/libmatio teams. See https://github.com/conda-forge/libnetcdf-feedstock/pull/1#issuecomment-222564591

**Comment by jakirkham** [comment]
> Can you point me to those here in this issue? I can only see failures due to version mismatch that should be fixed once this PR gets merged.

I'm trying to just catch up with what you have done an organize it a bit at this point.

**Comment by ocefpaf** [comment]
> I'm trying to just catch up with what you have done an organize it a bit at this point.

I just started! So I recommend you to wait a little bit :wink:

**Comment by jakirkham** [comment]
Only doing it so you can focus on adding PRs without recovering old ground. 😞

**Comment by ocefpaf** [comment]
> Only doing it so you can focus on adding PRs without recovering old ground. 😞

Glad you are on top of it. Just don't want you running some unnecessary errands before I am done.

**Comment by jakirkham** [comment]
Well, we should get organized at some point. If we wait until it is all done, it will just be a bigger mess. 😄

In any event, there is a table above with info about what packages have PRs for HDF5 and libnetcdf. Checkboxes don't work in tables so we will have to live with an `X` when they are merged. Hopefully, it makes everyone's life a little easier.

**Comment by kmuehlbauer** [comment]
@ocefpaf @jakirkham The hdf5 pinning at [libradolan](https://github.com/conda-forge/libradolan-feedstock/blob/master/recipe/meta.yaml#L24) can safely be removed, if all (!) netcdf-issues are sorted out. We had to put this in, because conda wasn't able to pull the right hdf5 at runtime.

**Comment by ocefpaf** [comment]
This is good to go and I will merge it tomorrow or the day after. The `bob.io.base`, `vigra`, and `libmatio` teams were pinged. I did not send PRs there as I don't know if it is in their interest to upgrade to the latest `hdf5` or not.

**Comment by jakirkham** [comment]
> The `bob.io.base`, `vigra`, and `libmatio` teams were pinged. I did not send PRs there as I don't know if it is in their interest to upgrade to the latest `hdf5` or not.

Could we please have pinning PRs for those as well if we are changing the stack? I would like us to remain compatible if we are making this change. Especially as I'm guessing this is being added to the pinning wiki.

**Comment by pelson** [comment]
> Could we please have pinning PRs for those as well if we are changing the stack?

I have the pinning code (for MNT PRs) on heroku now, so we can schedule that up for regular updates/maintenance. I think we are now in a position to start rolling out widescale conda-forge pinned version changes.

**Comment by jakirkham** [comment]
Cool, so what do we need to do?

**Comment by pelson** [comment]
> Cool, so what do we need to do?

Decide what needs pinning, and then turning on the scheduled task. It looks like many of the packages are already in hand with that excellent list of packages and associated PRs, but if we can automate the PR creation process it will mean nothing slips through the net.

1. What have we got lined up to upgrade in terms of pinned versions?
2. What is stopping us merging this PR now?

**Comment by jakirkham** [comment]
> What is stopping us merging this PR now?

See this list ( https://github.com/conda-forge/libnetcdf-feedstock/pull/1#issuecomment-222564591 ). Also, feel free to update with changes as they are made.

> What have we got lined up to upgrade in terms of pinned versions?

We need to switch everything over to HDF5 1.8.17*.

Unrelated, but it would be good if we update from Boost 1.60.* to 1.61.*, as well.

**Comment by pelson** [comment]
> Unrelated, but it would be good if we update from Boost 1.60.* to 1.61.*, as well.

Let's move that conversation to https://github.com/conda-forge/conda-forge.github.io/pull/156

**Comment by pelson** [comment]
> See this list ( #1 (comment) ). Also, feel free to update with changes as they are made.

So what isn't pinned correctly that would fail if we merged this now?

**Comment by pelson** [comment]
> So what isn't pinned correctly that would fail if we merged this now?

i.e. I'm having a slow day, so please talk me through it slowly :smile: 

I'm looking at libnetcdf's dependencies (not the packages that depend upon it) and can't see any that need to be fixed first AFAICT.

**Comment by jakirkham** [comment]
> > So what isn't pinned correctly that would fail if we merged this now?

> i.e. I'm having a slow day, so please talk me through it slowly :smile: 

Let's start with adding pinning PRs for `bob.io.base`, `vigra`, and `libmatio`.

**Comment by pelson** [comment]
> Let's start with adding pinning PRs for bob.io.base, vigra, and libmatio.

They *[are](https://github.com/conda-forge/bob.io.base-feedstock/blob/master/recipe/meta.yaml#L33)* pinned (to 1.8.15). I don't understand how we would *break* the environment if this PR were merged. Sure, you wouldn't be able to install this build of libnetcdf alongside those packages, but that is easy enough to fix - we just update those magic numbers. What am I missing?

Is there a broken build of one of the dependencies somewhere (even on defaults) which hasn't pinned things correctly and would therefore result in a broken installation?

**Comment by jakirkham** [comment]
> They *[are](https://github.com/conda-forge/bob.io.base-feedstock/blob/master/recipe/meta.yaml#L33)* pinned (to 1.8.15). I don't understand how we would *break* the environment if this PR were merged. Sure, you wouldn't be able to install this build of libnetcdf alongside those packages, but that is easy enough to fix - we just update those magic numbers. What am I missing that means there is a strict order in the way we do that?

I'm confused. Do we *not* want to try and use uniform pinnings across the stack now?

Besides my understanding is this would be trivial to do with that added code referenced above. Can we take this for a test drive?

**Comment by pelson** [comment]
> I'm confused. Do we not want to try and use uniform pinnings across the stack now?

Yes, completely. *Try* being a key word here. We can't and shouldn't hold an update back because a single package can't handle that updated version - we would end up crippling and never updating anything. Of course, there is some pragmatism needed too - if we ended up making it impossible to install a coherent environment with python and netcdf for instance, that would be a bad thing. If we had a situation where python, netcdf and vigra didn't install into a single environment, it isn't great, but it isn't a show-stopper either (note: I'm talking about conda raising an exception, not conda creating an environment that simply doesn't work)

It sounds to me like the only reason for holding back on this was to maintain consistency across the stack (and we were waiting on the PRs to tell us that every feedstock can accommodate the updated version), so I'm going to merge this in the morning (unless somebody does it before me). I'll ensure that PRs are available for all other feedstocks which have pins in the next 24 hours. IMO it is down to the maintainers of the respective feedstocks to ensure that those PRs are merged if their packages are to be installable with the wider conda-forge stack.

**Comment by jakirkham** [comment]
To be clear, I was only asking that the pinning PRs be made not necessarily merged before this is merged.

The important point that I would like to stress here is we need to keep in mind that this is a collaborative packaging effort. So, if we want to move the stack forward, we need to give some indication to others that we plan to do this. Adding pinning PRs are essential to that effort as they propagate that signal to the relevant parties.

**Comment by jakirkham** [comment]
As for the failures listed, it seems like many of them depend on HDF5 and libnetcdf or something else that depends on libnetcdf. It may be the case that merging this will just solve those failures. However, we can't really know until that happens.

**Comment by ocefpaf** [comment]
Thanks @pelson for firing up the bot! I am merging this so we can re-start the CIs in the dependencies and see how this goes.

**Pull request merged**